driver-core/src/main/java/com/datastax/driver/core/ControlConnection.java

```diff
@@ -434,81 +443,94 @@ void refreshNodeListAndTokenMap() {
+      return cluster.translateAddress(new InetSocketAddress(nativeAddress, nativePort));
+    } else {
+      // after CASSANDRA-9436, system.peers contains the following inet columns:
+      // - peer: this is actually broadcast_address
+      // - rpc_address: the address we are looking for (this corresponds to broadcast_rpc_address in
+      // the peer's cassandra yaml file;
+      //                if this setting if unset, it defaults to the value for rpc_address or
+      // rpc_interface)
+      // - preferred_ip: used by Ec2MultiRegionSnitch and GossipingPropertyFileSnitch, possibly
+      // others; contents unclear
+      InetAddress broadcastAddress = peersRow.getInet("peer");
+      InetAddress rpcAddress = peersRow.getInet("rpc_address");
+
+      if (broadcastAddress == null) {
+        return null;
+      } else if (broadcastAddress.equals(connectedHost.getAddress())
+          || (rpcAddress != null && rpcAddress.equals(connectedHost.getAddress()))) {
+        // Some DSE versions were inserting a line for the local node in peers (with mostly null
+        // values). This has been fixed, but if we
+        // detect that's the case, ignore it as it's not really a big deal.
+        logger.debug(
+            "System.peers on node {} has a line for itself. This is not normal but is a known problem of some DSE version. Ignoring the entry.",
+            connectedHost);
+        return null;
+      } else if (rpcAddress == null) {
+        return null;
+      } else if (rpcAddress.equals(bindAllAddress)) {
+        logger.warn(
+            "Found host with 0.0.0.0 as rpc_address, using broadcast_address ({}) to contact it instead. If this is incorrect you should avoid the use of 0.0.0.0 server side.",
+            broadcastAddress);
+        rpcAddress = broadcastAddress;
+      }
+      return cluster.translateAddress(rpcAddress);
     }
-    return cluster.translateAddress(rpcAddress);
   }
 
   private Row fetchNodeInfo(Host host, Connection c)
       throws ConnectionException, BusyConnectionException, ExecutionException,
           InterruptedException {
     boolean isConnectedHost = c.address.equals(host.getSocketAddress());
-    if (isConnectedHost || host.getBroadcastAddress() != null) {
+    if (isConnectedHost || host.getBroadcastSocketAddress() != null) {
+      String query;
+      if (isConnectedHost) {
+        query = SELECT_LOCAL;
+      } else {
+        InetSocketAddress broadcastAddress = host.getBroadcastSocketAddress();
+        query =
+            isPeersV2
+                ? SELECT_PEERS_V2
+                    + " WHERE peer='"
+                    + broadcastAddress.getAddress().getHostAddress()
+                    + "' AND peer_port="
+                    + broadcastAddress.getPort()
+                : SELECT_PEERS
+                    + " WHERE peer='"
+                    + broadcastAddress.getAddress().getHostAddress()
+                    + "'";
```

Comment:
We don't handle a missing `peers_v2` here, but that's fine: we do a `refreshNodeListAndTokenMap` every time the control connection (re)connects, and it happens before the new connection is published in `connectionRef`, so we know it will never race with this method.
So `isPeersV2` will always be set correctly at this point. Even in a mixed 3.x/4.0 cluster, if we initially connected to a 4.0 node, `isPeersV2` will downgrade correctly if we reconnect to a legacy node.